src/prefect/assets.py

```diff
@@ -0,0 +1,164 @@
+from __future__ import annotations
+
+import re
+from typing import Any, Callable, Sequence, TypeVar
+from uuid import UUID
+
+from pydantic import Field, field_validator
+from typing_extensions import ParamSpec
+
+from prefect import Task
+from prefect._internal.schemas.bases import PrefectBaseModel
+from prefect.client.schemas.objects import TaskRun
+from prefect.context import FlowRunContext
+from prefect.events import emit_event
+from prefect.states import State
+
+T = TypeVar("T")
+P = ParamSpec("P")
+R = TypeVar("R")
+
+
+URI_REGEX = re.compile(r"^[a-z0-9]+://")
+
+
+class Asset(PrefectBaseModel):
+    key: str
+    name: str | None = None
+    metadata: dict[str, str] = Field(default_factory=dict)
+
+    @field_validator("key")
+    @classmethod
+    def validate_key(cls, value: str) -> str:
+        if not URI_REGEX.match(value):
+            raise ValueError(
+                "Key must be a valid URI, e.g. storage://bucket/folder/asset.csv"
+            )
+        return value
+
+    def as_resource(self) -> dict[str, str]:
```

Comment:
I'm tempted to make these two helper methods private - I'm trying to keep public methods to a bare minimum of those methods we actually intend to document as solving a user-problem

Reply:
works for me

Reply:
Ah this breaks a pyright check:
``` 
error: "_as_related" is protected and used outside of the class in which it is declared (reportPrivateUsage)
```

let me just move them somewhere off the public facing class